src/proof/nova_lem.rs

```diff
@@ -0,0 +1,3348 @@
+
+    let commitment_size_hint1 = <SS1<F> as RelaxedR1CSSNARKTrait<G1<F>>>::commitment_key_floor();
+    let commitment_size_hint2 = <SS2<F> as RelaxedR1CSSNARKTrait<G2<F>>>::commitment_key_floor();
+
+    let pp = nova::PublicParams::setup(
+        &circuit_primary,
+        &circuit_secondary,
+        Some(commitment_size_hint1),
+        Some(commitment_size_hint2),
+    );
+    let (pk, vk) = CompressedSNARK::setup(&pp).unwrap();
+    PublicParams { pp, pk, vk }
+}
+
+impl<'a, F: CurveCycleEquipped> MultiFrame<'a, F> {
+    fn circuits(func: &'a Func, count: usize) -> (C1<'a, F>, C2<F>) {
+        (
+            MultiFrame::blank(func, count),
+            TrivialTestCircuit::default(),
+        )
+    }
+}
+
+/// A struct for the Nova prover that operates on field elements of type `F`.
+#[derive(Debug)]
+pub struct NovaProver<F: CurveCycleEquipped> {
+    // `reduction_count` specifies the number of small-step reductions are performed in each recursive step.
+    reduction_count: usize,
+    lang: Lang<F, DummyCoprocessor<F>>,
+    _p: PhantomData<F>,
+}
+
+impl<'a, F: CurveCycleEquipped> PublicParameters for PublicParams<'a, F>
+where
+    <<G1<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
+    <<G2<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
+{
+}
+
+impl<'a, F: CurveCycleEquipped> Prover<'a, '_, F, DummyCoprocessor<F>> for NovaProver<F>
+where
+    <<G1<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
+    <<G2<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
+{
+    type PublicParams = PublicParams<'a, F>;
+    fn new(reduction_count: usize, lang: Lang<F, DummyCoprocessor<F>>) -> Self {
+        NovaProver::<F> {
+            reduction_count,
+            lang,
+            _p: Default::default(),
+        }
+    }
+    fn reduction_count(&self) -> usize {
+        self.reduction_count
+    }
+
+    fn lang(&self) -> &Lang<F, DummyCoprocessor<F>> {
+        &self.lang
+    }
+}
+
+impl<F: CurveCycleEquipped> NovaProver<F>
+where
+    <<G1<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
+    <<G2<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
+{
+    /// Proves the computation given the public parameters, frames, and store.
+    pub fn prove<'a>(
+        &'a self,
+        func: &'a Func,
+        pp: &'a PublicParams<'_, F>,
+        frames: &[Frame<F>],
+        store: &'a mut Store<F>,
+    ) -> Result<(Proof<'_, F>, Vec<F>, Vec<F>, usize), ProofError> {
+        let z0 = store.to_vector(&frames.first().unwrap().input).unwrap();
+        let zi = store.to_vector(&frames.last().unwrap().output).unwrap();
+        let circuits = MultiFrame::from_frames(func, self.reduction_count(), frames, store);
+
+        let num_steps = circuits.len();
+        let proof =
+            Proof::prove_recursively(pp, store, &circuits, self.reduction_count, z0.clone())?;
+
+        Ok((proof, z0, zi, num_steps))
+    }
```

Comment:
Why is this redefining every struct in the nova prover in order to define this? You'll have to re-add the very same generic parameter once you do put in coprocessor support. You can simply instantiate for the coprocessor you do happen to work with. In concrete terms, you can implement:
```rust
impl<F: CurveCycleEquipped> NovaProver<F, DummyCoprocessor<F>>
where
    <<G1<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
    <<G2<F> as Group>::Scalar as ff::PrimeField>::Repr: Abomonation,
{
    <your logic here>
}
```